flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/FlinkDistribution.java

```diff
@@ -123,20 +126,28 @@ public void afterTestSuccess() {
 
 	@Override
 	public void afterTestFailure() {
-		logBackupDir.ifPresent(backupLocation -> {
+		if (logBackupDir != null) {
 			final UUID id = UUID.randomUUID();
-			LOG.info("Backing up logs to {}/{}.", backupLocation, id);
+			LOG.info("Backing up logs to {}/{}.", logBackupDir, id);
 			try {
-				Files.createDirectories(backupLocation);
-				FileUtils.copyDirectory(log.toFile(), backupLocation.resolve(id.toString()).toFile());
+				Files.createDirectories(logBackupDir);
+				FileUtils.copyDirectory(log.toFile(), logBackupDir.resolve(id.toString()).toFile());
 			} catch (IOException e) {
 				LOG.warn("An error occurred while backing up logs.", e);
 			}
-		});
-
+		}
 		afterTestSuccess();
 	}
 
+	/**
+	 * Read the value of `rest.port` part in FLINK_DIST_DIR/conf/flink-conf.yaml.
+	 *
+	 * @return the rest port which standalone Flink cluster will listen.
+	 */
+	public int getRestPort() {
+		return defaultConfig.getInteger("rest.port", 8081);
```

Comment:
Where would we set the port to 0 ? Searched the rest.port in flink project, seems no place will do this...

Reply:
As I said, tests will _eventually_ set this to 0, or more accurately, the FlinkResource/FlinkDistribution implementation will do so to avoid port conflicts.
Just likey all unit tests that rely on a `MiniCluster[Resource]` that have the port set to 0.

Currently, the test in this PR relies on the default port (8081), purely because the existing e2e tests make the same assumption. (As such there was no immediate need to fix it)
This is _obviously_ problematic in the long run, but fine for now.

In contrast, relying on a setting in the configuration is just as problematic (in the case of 0), but looks reasonable at a glance, making it a worse approach.

I'd just replace this right her with a hard-coded `8081`, to make it very obvious that this bit is problematic.

The _long-term_ solution is to set this port to 0 and parse the actually used port from the log file.

Reply:
Well, thanks for your careful explantation, really helpful. 
For the long-term solution,   if use the random port, then we can only write a small method to sweep the ports to detect which one is the rest.port ?  Or  the MiniCluster need to pre-check the port one by one, if available then choose to be its rest.port and bind & listen ( As far as I know , the HBase mini cluster is using this way) , and after that the minicluster write the port into flink-config.yaml  ...  seems the second way will be better  I mean..

Reply:
We can't rely on port sweeps, since the goal is to be able to run these tests in parallel; in that case multiple flink clusters (and thus REST APIs) are running.

The flink cluster should not be modifying the flink.conf.yaml because this is a user-controlled file. Similarly, creating _some_ file containing the port pollutes the distribution, which should be avoided in production settings, which we are testing.
This would only be acceptable if the file were only created for the test, but that implies test-specific logic in production code that we again don't want.

I think that running a regex against the logs is likely the easiest and most reliable solution, even if it is prone to breakage if the log message changes.

Reply:
Fine,  let me revert to use 8081 firstly.  Will file a separate issue to improve the port management. Thanks much.